### ISSUE TITLE
helm: PSPs are now deprecated, investigate using Pod Security Admission

### ISSUE THREAD

**Issue by hugoShaka** [gravitational/teleport#20088]
### What would you like Teleport to do?

It should run by default under the `restricted` security standard if possible.

### What problem does this solve?

As an admin deploying Teleport I expect the application to be deployed with the strictest security boundaries possible.

### If a workaround exists, please include it.

A user would set security labels manually, [but this is not supported by the chart](https://github.com/gravitational/teleport/issues/19381).

**Comment by zmb3** [comment]
I believe this is also what was preventing me from merging #12760

**Comment by hugoShaka** [comment]
> I believe this is also what was preventing me from merging #12760

We can't really remove PSPs, and Helm is always declaring them supported for compatibility reasons. I think the easiest workaround would be to set `psp.enabled: false` on the lint commands

**Comment by klizhentas** [comment]
PSP will be removed in 12.4, so we should remove them too I guess

**Comment by hugoShaka** [comment]
The chart already detects clusters not supporting PSPs and disables the feature. It installs properly in 1.25 and 1.26 clusters. However I'm not sure the upgrade is seamless, I think Helm won't be happy when trying to delete the resource through an unsupported API.

We can disable PSPs by default for all versions, but this will be a breaking change for <1.24 users. I don't think it's a big deal for 1.23+ users as Pod Security Admission should be supported and we might be able to run in `restricted` mode. Disabling PSPs without replacement (1.22 and before) would represent a security downgrade and a more questionable change IMO.

I would be more comfortable with dynamically disabling PSPs when running under 1.23/1.24, rather than removing them entirely, including for Kubernetes versions still supported. Also, a non-negligible portion of our users seems to run on *old* Kubernetes versions.

**Comment by klizhentas** [comment]
@hugoShaka what happens when customers upgrade their clusters with old versions of helm charts running, would it break for them?

**Comment by hugoShaka** [comment]
I have not been able to test, as most local setups don't support upgrading. I suspect Helm is not smart enough to handle the PSP removal properly. Upgrading the Kubernetes will work, but the next Helm `upgrade` will try to remove PSPs invoking the deprecated APIs and return an error. The deployment will never be affected, but Helm's state will end up out of sync with the cluster. Last time I checked, recovering from such a state without downtime is extremely painful.

The condition for the Helm upgrade to work is to have run at least 1 `helm upgrade teleport-cluster` on a Kubernetes version supporting PSPs with the chart deleting PSPs. If we disable PSPs on 1.23 and 1.24 this should give enough time for users to upgrade at least once.

I don't see another way to handle this, Helm should be able to deal with this, but it seems they chose not to.

**Comment by hugoShaka** [comment]
## The problems

After playing with PSAs and upgrades I see 2 issues:
- Upgrading from 1.X to 1.25 will break the Helm state (see previous message)
- PSAs cannot be configured from Helm (explanation later)

PSA applies security levels (PSS) for a whole namespace. There are 3 levels (privileged, baseline, restricted), and a label on the namespace controls the PSA behaviour. Helm3 releases have to live in a namespace (it has to pre-exist the Helm release, it can be bootstrapped with a CLI flag but is not part of the release). Helm cannot edit resources it doesn't own. Because Helm cannot own the namespace, Helm cannot configure PSA.

tl;dr, we have no way of enforcing the PSS, we can only instruct users to set it through documentation and release notes.

## Current state

I tested PSS levels and `teleport-cluster` runs fine on "baseline" level, however "restricted" will require a bit more work, especially because we're running as root in the images. We should use "baseline" until we can properly support "restricted" (the item has been submitted as a potential Q1 goal).

`teleport-kube-agent` might even work properly under restricted PSS.

## The plan

After discussing with @r0mant we ended up with the following plan:
- For v10 and v11, keep the PSPs enabled and add warning in the `Notes.txt`, in the docs and the release notes telling users to disable PSPs and enable PSA before upgrading to 1.25
- For v12:
  - For <1.23 deploy PSPs and add a warning to set PSA
  - For 1.23 and 1.24 don't deploy PSPs and add a warning to set PSA (this is a breaking change)
  - For 1.25+ add a warning to set PSA
- For v13:
  - Don't deploy PSPs and remove the warning?
  
I'll send a PR disabling PSPs, updating the guides and adding the warning today so we can be ready for v12.
